x/delayedack/ibc_middleware.go

```diff
@@ -0,0 +1,244 @@
+}
+
+/* ------------------------------- ICS4Wrapper ------------------------------ */
+
+// SendPacket implements the ICS4 Wrapper interface
+func (im IBCMiddleware) SendPacket(
+	ctx sdk.Context,
+	chanCap *capabilitytypes.Capability,
+	sourcePort string,
+	sourceChannel string,
+	timeoutHeight clienttypes.Height,
+	timeoutTimestamp uint64,
+	data []byte,
+) (sequence uint64, err error) {
+	return im.keeper.SendPacket(ctx, chanCap, sourcePort, sourceChannel, timeoutHeight, timeoutTimestamp, data)
+}
+
+// WriteAcknowledgement implements the ICS4 Wrapper interface
+func (im IBCMiddleware) WriteAcknowledgement(
+	ctx sdk.Context,
+	chanCap *capabilitytypes.Capability,
+	packet exported.PacketI,
+	ack exported.Acknowledgement,
+) error {
+	return im.keeper.WriteAcknowledgement(ctx, chanCap, packet, ack)
+}
+
+// GetAppVersion returns the application version of the underlying application
+func (im IBCMiddleware) GetAppVersion(ctx sdk.Context, portID, channelID string) (string, bool) {
+	return im.keeper.GetAppVersion(ctx, portID, channelID)
+}
+
+/* ------------------------------- Custom Logic ------------------------------ */
+
+// AfterStateFinalized implements the RollappHooks interface
+func (im IBCMiddleware) AfterStateFinalized(ctx sdk.Context, rollappID string, stateInfo *rollapptypes.StateInfo) error {
+	// Finalize the packets for the rollapp at the given height
+	stateEndHeight := stateInfo.StartHeight + stateInfo.NumBlocks - 1
+	if err := im.FinalizeRollappPackets(ctx, rollappID, stateEndHeight); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// FinalizeRollappPackets finalizes the packets for the given rollapp until the given height which is
+// the end height of the latest finalized state
+func (im IBCMiddleware) FinalizeRollappPackets(ctx sdk.Context, rollappID string, stateEndHeight uint64) error {
+	logger := ctx.Logger().With("module", "DelayedAckMiddleware")
+	// Get the packets for the rollapp until height
+	rollappPendingPackets := im.keeper.ListRollappPendingPackets(ctx, rollappID, stateEndHeight)
+	logger.Debug("Finalizing IBC rollapp packets", "rollappID", rollappID, "state end height", stateEndHeight, "num packets", len(rollappPendingPackets))
+	for _, rollappPacket := range rollappPendingPackets {
+		// Update the packet status
+		im.keeper.UpdateRollappPacketStatus(ctx, rollappID, rollappPacket, types.RollappPacket_ACCEPTED)
+		// Call the onRecvPacket callback for each packet
+		ack := im.app.OnRecvPacket(ctx, *rollappPacket.Packet, rollappPacket.Relayer)
+		// Write the acknowledgement to the chain only if it is synchronous
+		if ack != nil {
+			_, chanCap, err := im.keeper.LookupModuleByChannel(ctx, rollappPacket.Packet.DestinationPort, rollappPacket.Packet.DestinationChannel)
+			if err != nil {
+				logger.Error("Error looking up module by channel", "rollappID", rollappID, "error", err.Error())
+				return err
+			}
+			err = im.keeper.WriteAcknowledgement(ctx, chanCap, rollappPacket.Packet, ack)
```

Comment:
why it's needed? isn't u just need to run `im.app.OnRecvPacket` as we skipped it previously in the middleware?

Reply:
If the `OnRecvPacket` returns an ack (i.e sync acknowledgement) then you don’t need it, but if it returns nil (async ack), then you need to call it by yourself once the ack is received.